File: 1818-maximum-score-from-removing-substrings/maximum-score-from-removing-substrings.cpp
class Solution {
public:
    int maximumGain(string s, int x, int y) {
        int primary_bonus = max(x, y);
        int secondary_bonus = min(x, y);
        vector<char> primary;
        vector<char> secondary;

        if (x >= y){
            primary = {'a', 'b'};
            secondary = {'b', 'a'};
        } else{
            primary = {'b', 'a'};
            secondary = {'a', 'b'};
        }

        stack<pair<char, int>> prev = {};
        int total = 0;
        for (int ind = 0; ind < s.size(); ind++){
            // match all primary first
            if (!prev.empty() && prev.top().first == primary[0] && s[ind] == primary[1]){
                s[prev.top().second] = '.';
                s[ind] = '.';
                prev.pop();
                total += primary_bonus;
            }else{
                prev.push({s[ind], ind});
            }
        }

        stack<char> prev2 = {};
        for (int ind = 0; ind < s.size(); ind++){
            // now match secondary
            if (s[ind] == '.') continue;

            if (!prev2.empty() && prev2.top() == secondary[0] && s[ind] == secondary[1]){
                prev2.pop();
                total += secondary_bonus;
            }else{
                prev2.push(s[ind]);
            }
        }

        return total;
    }
};

File: 1818-maximum-score-from-removing-substrings/maximum-score-from-removing-substrings.py
class Solution:
    def stripPattern(self, s, pattern, reward):
        result = []
        total = 0
        
        for c in s:
            result.append(c)

            while len(result) >= 2 and result[-2] == pattern[0] and result[-1] == pattern[1]:
                result.pop()
                result.pop()
                total += reward

        return result, total

    def maximumGain(self, s: str, x: int, y: int) -> int:
        # Suppose x is more valuable: just take all the ab you can get in first pass.
        # then you will just have at most bbbbbaaaaa, which you can take as many as needed
        # To do one pass: have a stack for all chars so far, remove as needed
        # Q: Will taking ab ever result in us not taking [ba] x2. No! Because baba -> ba so we still get one of them
        patternOrder = ["ab", "ba"]
        if y > x:
            patternOrder = ["ba", "ab"]
        
        res1, tot1 = self.stripPattern(s, patternOrder[0], max(x, y))
        res2, tot2 = self.stripPattern(res1, patternOrder[1], min(x, y))
        return tot1 + tot2

        